### PR TITLE
Master parallel hoist

### PR DESCRIPTION
**Pull request by weiya711**
Parallel workspace hoisting code for CPUThread based on the master branch. This has been tested for the spgemm.scheduling_eval test. It should work with both -DOPENMP=ON and -DOPENMP=OFF.

The files changed were mainly:
1. Lowering changes to support parallel hoisting in `src/lower/lowerer_impl.cpp`
2. Header changes in `src/codegen/codegen_c.cpp`
3. `Add::make(Expr a, Expr b, Datatype type)` changes in index_notation to support address arithmetic on boolean pointers without throwing a taco assertion error. It now checks if neither input operand is a boolean and if neither is a variable pointer
4. Add in an error message printout in `src/codegen/module.cpp` if `dlopen(...)` fails using `dlerror()`

### CODE REVIEW

File: src/lower/lowerer_impl.cpp [hunk 1]
```diff
@@ -697,11 +697,15 @@ Stmt LowererImpl::lowerForall(Forall forall)
...
                                         reducedAccesses);
 
-  // Emit temporary initialization if forall is sequential and leads to a where statement
+  // Emit temporary initialization if forall is sequential or parallelized by cpu threads and leads to a where statement
+  // This is for workspace hoisting by 1-level
   vector<Stmt> temporaryValuesInitFree = {Stmt(), Stmt()};
   auto temp = temporaryInitialization.find(forall);
   if (temp != temporaryInitialization.end() && forall.getParallelUnit() == ParallelUnit::NotParallel && !isScalar(temp->second.getTemporary().getType()))
     temporaryValuesInitFree = codeToInitializeTemporary(temp->second);
+  else if (temp != temporaryInitialization.end() && forall.getParallelUnit() == ParallelUnit::CPUThread && !isScalar(temp->second.getTemporary().getType())) {
```
Comment (stephenchouca) [inline src/lower/lowerer_impl.cpp:706]: Some of these lines seem a bit long. Ideally, no line should be more than 80 characters long.

File: test/tests-scheduling.cpp [hunk 2]
```diff
@@ -845,9 +845,9 @@ TEST(scheduling_eval_test, spmv_fuse) {
           .parallelize(block, ParallelUnit::GPUBlock, OutputRaceStrategy::IgnoreRaces)
           .parallelize(warp, ParallelUnit::GPUWarp, OutputRaceStrategy::Atomics)
           .parallelize(thread, ParallelUnit::GPUThread, OutputRaceStrategy::Atomics);
-//  ir::CodeGen_CUDA codegen = ir::CodeGen_CUDA(cout, ir::CodeGen_CUDA::ImplementationGen);
-//  ir::Stmt compute = lower(stmt, "compute",  false, true);
-//  codegen.print(compute);
+  ir::CodeGen_CUDA codegen = ir::CodeGen_CUDA(cout, ir::CodeGen_CUDA::ImplementationGen);
```
Comment (stephenchouca) [inline test/tests-scheduling.cpp:848]: What is this for?

Comment (weiya711) [reply test/tests-scheduling.cpp:848]: This was Ryan's method of testing what the CUDA code generation would print out in terminal when you run a googletest ctest. I can comment it back out (which I should probably do).

File: src/lower/lowerer_impl.cpp (hunk 1 above)
Comment (weiya711) [reply src/lower/lowerer_impl.cpp]: I reduced the line sizes for lines: 700, 706, 2123, and 2126 in `src/lower/lowerer_impl.cpp`. 

All of the other lines in my changes seem to be within the stylistic limit of the github screen (80-100 characters).

**Pull request merged**